pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -187,40 +191,49 @@ def restore_loops(self) -> None:
-        )
-        n_accum = 1 if self.trainer.accumulate_grad_batches is None else self.trainer.accumulate_grad_batches
-        expected_steps = self.trainer.num_training_batches / n_accum
-        if self.trainer.num_training_batches != 0 and self.trainer.global_step % expected_steps > 1:
-            rank_zero_warn(
-                "You're resuming from a checkpoint that ended mid-epoch."
-                " Training will start from the beginning of the next epoch."
-                " This can cause unreliable results if further training is done,"
-                " consider using an end of epoch checkpoint."
+        if self.trainer.state.fn == TrainerFn.FITTING:
+            self.trainer.fit_loop.global_step = self._loaded_checkpoint["global_step"]
+            self.trainer.fit_loop.current_epoch = self._loaded_checkpoint["epoch"]
+
+            # crash if max_epochs is lower then the current epoch from the checkpoint
+            if (
+                FitLoop._is_max_limit_enabled(self.trainer.max_epochs)
+                and self.trainer.current_epoch > self.trainer.max_epochs
+            ):
+                raise MisconfigurationException(
+                    f"You restored a checkpoint with current_epoch={self.trainer.current_epoch},"
+                    f" but you have set Trainer(max_epochs={self.trainer.max_epochs})."
+                )
+
+            # Division deals with global step stepping once per accumulated batch
+            # Inequality deals with different global step for odd vs even num_training_batches
+            self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(
+                self.trainer.current_epoch
             )
+            n_accum = 1 if self.trainer.accumulate_grad_batches is None else self.trainer.accumulate_grad_batches
+            expected_steps = self.trainer.num_training_batches / n_accum
+            if self.trainer.num_training_batches != 0 and self.trainer.global_step % expected_steps > 1:
+                rank_zero_warn(
+                    "You're resuming from a checkpoint that ended mid-epoch."
+                    " Training will start from the beginning of the next epoch."
+                    " This can cause unreliable results if further training is done,"
+                    " consider using an end of epoch checkpoint."
+                )
 
         state_dict = self._loaded_checkpoint.get("loops")
-        if state_dict:
-            self.trainer.fit_loop.load_state_dict(state_dict["fit_loop"])
-            self.trainer.validate_loop.load_state_dict(state_dict["validate_loop"])
-            self.trainer.test_loop.load_state_dict(state_dict["test_loop"])
-            self.trainer.predict_loop.load_state_dict(state_dict["predict_loop"])
+
+        if state_dict is not None:
+            if self.trainer.state.fn == TrainerFn.FITTING:
+                self.trainer.fit_loop.load_state_dict(state_dict.get("fit_loop"))
```

Comment:
not sure why we're using `state_dict.get(...)` here instead of direct index access. this would avoid the optional typing in the loops load state dict. shouldn't it be enforced that the state dict has a fit_loop entry if we're resuming for fitting?

Reply:
yep.. actually wanted to confirm this. Should it be enforced or not? Ideally, it should be but just wanted other's thoughts on this.